### PR TITLE
race condition when running solana-program-test

### PR DESCRIPTION
**Pull request by mschneider**
#### Problem

The banks-server transaction execution used by solana-program-test hits a race condition when executing large test suites. Here's [an example](https://github.com/blockworks-foundation/mango-v4/actions/runs/4966814314/) of such a run, which only succeeded on the 3rd attempt.

```
2023-05-13T12:26:36.4991091Z thread 'solBankForksCli' panicked at 'assertion failed: !self.freeze_started()', /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/solana-runtime-1.14.17/src/bank.rs:6201:9
2023-05-13T12:26:36.4991704Z note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
2023-05-13T12:26:36.4992796Z thread 'cases::test_perp_settle::test_perp_settle_pnl' panicked at 'called `Result::unwrap()` on an `Err` value: "SendError(..)"', /home/runner/.cargo/registry/src/github.com-1ecc6299db9ec823/solana-banks-server-1.14.17/src/banks_server.rs:333:44
2023-05-13T12:26:36.4994208Z thread 'cases::test_perp_settle::test_perp_settle_pnl' panicked at 'called `Result::unwrap()` on an `Err` value: IoError(Custom { kind: Other, error: "the request exceeded its deadline" })', programs/mango-v4/tests/program_test/mango_client.rs:391:6
```

The issue here is that the highest slot bank is sometimes frozen by the time `try_process_transactions` is called.

#### Summary of Changes

Unlike BankingStage, BanksServer does not stop to schedule shortly before the end of the slot. I didn't port this behavior as a slow block shouldn't really hurt a solana-program-test run, but instead acquired the bank's freeze lock for the duration of the processing.

This problem is very difficult to reproduce, so I didn't add a test (yet). Would be grateful for ideas how to test it using a unit test as this kind of issue can re-appear anytime given the complexity of locks hierarchies around the Bank.

### CODE REVIEW

File: banks-server/src/banks_server.rs
Comment (mvines) [inline banks-server/src/banks_server.rs:94]: nit: seems more readable if an implicit `continue` is used
```rust
                if *lock == Hash::default() {
                    let _ = bank.try_process_transactions(transactions.iter());
                    // break out of inner loop and release bank freeze lock
                    break;
                }
                // retry reading working bank if bank is already frozen
```

File: banks-server/src/banks_server.rs
Comment (mvines) [inline banks-server/src/banks_server.rs:101]: nit: seems more readable if an implicit `continue` is used
```suggestion
                if *lock == Hash::default() {
                    let _ = bank.try_process_transactions(transactions.iter());
                    // break out of inner loop and release bank freeze lock
                    break;
                }
                // retry reading working bank if bank is already frozen
```

File: banks-server/src/banks_server.rs
Comment (apfitzge) [inline banks-server/src/banks_server.rs:92]: this isn't quite how BankingStage does it. We only hold the freeze lock to prevent freezing after execution, while we are recording/committing.

Ref: https://github.com/solana-labs/solana/blob/master/core/src/banking_stage/consumer.rs#L524-L579

Comment (apfitzge) [reply banks-server/src/banks_server.rs:92]: Because we don't take the lock anywhere inside `try_process_transactions`, I think it's better to grab and hold the lock. @mvines suggestion seems clean (can also remove the loop label)

**Review by mvines — APPROVED** [review]
lgtm, @apfitzge if you agree can you please mash the green merge button

**Review by apfitzge — APPROVED** [review]
lgtm!

**Pull request merged**
